dsp-value-objects/src/main/scala/dsp/valueobjects/Project.scala

```diff
@@ -0,0 +1,184 @@
+  object ProjectDescription { self =>
+    def make(value: Seq[V2.StringLiteralV2]): Validation[Throwable, ProjectDescription] =
+      if (value.isEmpty) {
+        Validation.fail(V2.BadRequestException(ProjectErrorMessages.ProjectDescriptionMissing))
+      } else {
+        Validation.succeed(new ProjectDescription(value) {})
+      }
+
+    def make(value: Option[Seq[V2.StringLiteralV2]]): Validation[Throwable, Option[ProjectDescription]] =
+      value match {
+        case Some(v) => self.make(v).map(Some(_))
+        case None    => Validation.succeed(None)
+      }
+  }
+
+  /**
+   * Project Keywords value object.
+   */
+  sealed abstract case class Keywords private (value: Seq[String])
+  object Keywords { self =>
+    def make(value: Seq[String]): Validation[Throwable, Keywords] =
+      if (value.isEmpty) {
+        Validation.fail(V2.BadRequestException(ProjectErrorMessages.KeywordsMissing))
+      } else {
+        Validation.succeed(new Keywords(value) {})
+      }
+
+    def make(value: Option[Seq[String]]): Validation[Throwable, Option[Keywords]] =
+      value match {
+        case Some(v) => self.make(v).map(Some(_))
+        case None    => Validation.succeed(None)
+      }
+  }
+
+  /**
+   * Project Logo value object.
+   */
+  sealed abstract case class Logo private (value: String)
+  object Logo { self =>
+    def make(value: String): Validation[Throwable, Logo] =
+      if (value.isEmpty) {
+        Validation.fail(V2.BadRequestException(ProjectErrorMessages.LogoMissing))
+      } else {
+        Validation.succeed(new Logo(value) {})
+      }
+    def make(value: Option[String]): Validation[Throwable, Option[Logo]] =
+      value match {
+        case Some(v) => self.make(v).map(Some(_))
+        case None    => Validation.succeed(None)
+      }
+  }
+
+  /**
+   * ProjectSelfjoin value object.
+   */
+  sealed abstract case class ProjectSelfJoin private (value: Boolean)
+  object ProjectSelfJoin { self =>
+    def make(value: Boolean): Validation[Throwable, ProjectSelfJoin] =
+      Validation.succeed(new ProjectSelfJoin(value) {})
+
+    def make(value: Option[Boolean]): Validation[Throwable, Option[ProjectSelfJoin]] =
+      value match {
+        case Some(v) => self.make(v).map(Some(_))
+        case None    => Validation.succeed(None)
+      }
+  }
+
+  /**
+   * ProjectStatus value object.
+   */
+  sealed abstract case class ProjectStatus private (value: Boolean)
+  object ProjectStatus { self =>
+    def make(value: Boolean): Validation[Throwable, ProjectStatus] =
+      Validation.succeed(new ProjectStatus(value) {})
+
+    def make(value: Option[Boolean]): Validation[Throwable, Option[ProjectStatus]] =
+      value match {
+        case Some(v) => self.make(v).map(Some(_))
+        case None    => Validation.succeed(None)
+      }
+  }
+}
+
+object ProjectErrorMessages {
+  val ShortcodeMissing          = "Shortcode cannot be empty."
+  val ShortcodeInvalid          = "Shortcode is invalid."
```

Comment:
in these instances it would be a much more useful error message if it did contain the passed shortcode (and equally in most of those error messages.  
So maybe it should be something like:

```scala
val ShortcodeInvalid = (sc: String) => s"Shortcode '$sc' is invalid."
```

And then use it like:

```scala
throw V2.BadRequestException(ProjectErrorMessages.ShortcodeInvalid(shortcode))
```

Reply:
but maybe this is also not necessary if we use ZIO Validation?

Reply:
Subject to discuss.